In [178]:
import os
os.getcwd()

'C:\\Users\\DELL\\Capstone projects - Practised'

In [179]:
#Importing all the necessary libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

%matplotlib inline
import warnings
warnings.filterwarnings("ignore")

In [180]:
mov = pd.read_csv(r"C:\Users\DELL\CAPSTONE PROJECTS\Recommendation System\movies.csv")
cred = pd.read_csv(r"C:\Users\DELL\CAPSTONE PROJECTS\Recommendation System\credits.csv")

In [181]:
mov.head(3)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26   880674609   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466

In [182]:
cred.head(3)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...

In [183]:
#Shape of the dataset
mov.shape

(4803, 20)

In [184]:
cred.shape

(4803, 4)

In [185]:
#Finding the info of the dataset

mov.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4803 non-null   int64  
 1   genres                4803 non-null   object 
 2   homepage              1712 non-null   object 
 3   id                    4803 non-null   int64  
 4   keywords              4803 non-null   object 
 5   original_language     4803 non-null   object 
 6   original_title        4803 non-null   object 
 7   overview              4800 non-null   object 
 8   popularity            4803 non-null   float64
 9   production_companies  4803 non-null   object 
 10  production_countries  4803 non-null   object 
 11  release_date          4802 non-null   object 
 12  revenue               4803 non-null   int64  
 13  runtime               4801 non-null   float64
 14  spoken_languages      4803 non-null   object 
 15  status               

In [186]:
cred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   movie_id  4803 non-null   int64 
 1   title     4803 non-null   object
 2   cast      4803 non-null   object
 3   crew      4803 non-null   object
dtypes: int64(1), object(3)
memory usage: 150.2+ KB


In [187]:
#Checking for unique values in title

print(mov["title"].nunique())
print()
print(cred["title"].nunique())

#Unique values are the same for both the cred and mov dataframes under the col title

4800

4800


In [188]:
#Merging the columns of both the dataframes on title
movies = pd.merge(left=mov, right=cred,on="title",how="inner")
movies.head(2)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   

                                       homepage     id  \
0                   http://www.avatarmovie.com/  19995   
1  http://disney.go.com/disneypictures/pirates/    285   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   

                                production_companies  ... runtime  \
0  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...  ...   169.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   

                                      title vote_average vote_count movie_id  \
0                                    Avatar          7.2      11800    19995   
1  Pirates of the Caribbean: At World's End          6.9       4500      285   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  

[2 rows x 23 columns]

In [189]:
movies.shape
#There are some movie titles which have been repeated hence 4809

(4809, 23)

In [190]:
movies["title"].nunique()

4800

In [191]:
movies["title"].value_counts()
#The first 3 have been repeated extra 3 times hence the shape of movies dataframe is 4809 rows and 23 cols

The Host                 4
Batman                   4
Out of the Blue          4
Avatar                   1
The Girl on the Train    1
                        ..
Step Up 3D               1
Secondhand Lions         1
The Age of Adaline       1
Drag Me to Hell          1
My Date with Drew        1
Name: title, Length: 4800, dtype: int64

In [192]:
movies.columns

Index(['budget', 'genres', 'homepage', 'id', 'keywords', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'vote_average',
       'vote_count', 'movie_id', 'cast', 'crew'],
      dtype='object')

In [193]:
#Checking for null values
movies.isna().mean()*100
#Drop the homepage col

budget                   0.000000
genres                   0.000000
homepage                64.379289
id                       0.000000
keywords                 0.000000
original_language        0.000000
original_title           0.000000
overview                 0.062383
popularity               0.000000
production_companies     0.000000
production_countries     0.000000
release_date             0.020794
revenue                  0.000000
runtime                  0.041589
spoken_languages         0.000000
status                   0.000000
tagline                 17.550426
title                    0.000000
vote_average             0.000000
vote_count               0.000000
movie_id                 0.000000
cast                     0.000000
crew                     0.000000
dtype: float64

In [194]:
movies["original_language"].value_counts()/len(movies)*100
#94% consists of english language so you can actually drop the col

en    93.782491
fr     1.455604
es     0.665419
zh     0.561447
de     0.561447
hi     0.395093
ja     0.332710
it     0.291121
ko     0.249532
cn     0.249532
ru     0.228738
pt     0.187149
da     0.145560
sv     0.103972
nl     0.083177
fa     0.083177
th     0.062383
he     0.062383
ta     0.041589
cs     0.041589
ro     0.041589
id     0.041589
ar     0.041589
vi     0.020794
sl     0.020794
ps     0.020794
no     0.020794
ky     0.020794
hu     0.020794
pl     0.020794
af     0.020794
nb     0.020794
tr     0.020794
is     0.020794
xx     0.020794
te     0.020794
el     0.020794
Name: original_language, dtype: float64

In [195]:
movies["status"].value_counts()

Released           4801
Rumored               5
Post Production       3
Name: status, dtype: int64

In [196]:
movies["release_date"].describe()

count           4808
unique          3280
top       2006-01-01
freq              10
Name: release_date, dtype: object

In [197]:
#Removing the cols which do not make much sense for a content based recommendation movie system

movies_1 = movies.copy()
cols = ['budget','homepage', 'id','original_language','original_title','popularity', 'production_companies','production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline','vote_average','vote_count']

movies_1 = movies_1.drop(columns=cols,axis=1)
movies_1.head(2)


genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   

                                      title  movie_id  \
0                                    Avatar     19995   
1  Pirates of the Caribbean: At World's End       285   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...

In [198]:
movies_1.isna().mean()*100

genres      0.000000
keywords    0.000000
overview    0.062383
title       0.000000
movie_id    0.000000
cast        0.000000
crew        0.000000
dtype: float64

In [199]:
movies_1["genres"][0]

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

In [200]:
movies_1["keywords"][0]

'[{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "space war"}, {"id": 3388, "name": "space colony"}, {"id": 3679, "name": "society"}, {"id": 3801, "name": "space travel"}, {"id": 9685, "name": "futuristic"}, {"id": 9840, "name": "romance"}, {"id": 9882, "name": "space"}, {"id": 9951, "name": "alien"}, {"id": 10148, "name": "tribe"}, {"id": 10158, "name": "alien planet"}, {"id": 10987, "name": "cgi"}, {"id": 11399, "name": "marine"}, {"id": 13065, "name": "soldier"}, {"id": 14643, "name": "battle"}, {"id": 14720, "name": "love affair"}, {"id": 165431, "name": "anti war"}, {"id": 193554, "name": "power relations"}, {"id": 206690, "name": "mind and soul"}, {"id": 209714, "name": "3d"}]'

We need to extract the value of names: from each row of the genres and keywords cols

In [201]:
#Using ast librabry extracting the values of name in genres and keywords column
#ast library is used on dictionary key:value pairs

import ast
ast.literal_eval(movies_1["genres"][0])

[{'id': 28, 'name': 'Action'},
 {'id': 12, 'name': 'Adventure'},
 {'id': 14, 'name': 'Fantasy'},
 {'id': 878, 'name': 'Science Fiction'}]

In [202]:
#Creating a function to extract the values of name
def abstract (text):
    lst =[]
    
    for i in ast.literal_eval(text):#The text will be row and then i will be each dictionary in that row
        lst.append(i["name"])
        
    return lst

In [203]:
movies_1["genres"] = movies_1["genres"].apply(abstract)
movies_1["keywords"] = movies_1["keywords"].apply(abstract)

In [204]:
movies_1.head(3)

genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   

                                      title  movie_id  \
0                                    Avatar     19995   
1  Pirates of the Caribbean: At World's End       285   
2                                   Spectre    206647   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...

In [205]:
movies_1["cast"][0]

'[{"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}, {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}, {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}, {"cast_id": 4, "character": "Col. Quaritch", "credit_id": "52fe48009251416c750ac9cf", "gender": 2, "id": 32747, "name": "Stephen Lang", "order": 3}, {"cast_id": 5, "character": "Trudy Chacon", "credit_id": "52fe48009251416c750ac9d3", "gender": 1, "id": 17647, "name": "Michelle Rodriguez", "order": 4}, {"cast_id": 8, "character": "Selfridge", "credit_id": "52fe48009251416c750ac9e1", "gender": 2, "id": 1771, "name": "Giovanni Ribisi", "order": 5}, {"cast_id": 7, "character": "Norm Spellman", "credit_id": "52fe48009251416c750ac9dd", "gender": 

In [206]:
#Fetching the first 3 names of the cast from the data

def fetch (txt):
    
    lst_1=[]
    
    count=0
    for i in ast.literal_eval(txt):
        if count<3:
            lst_1.append(i["name"])
            count+=1
    return lst_1
        

In [207]:
movies_1["cast"]=movies_1["cast"].apply(fetch)

movies_1.head()

genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                      title  movie_id  \
0                                    Avatar     19995   
1  Pirates of the Caribbean: At World's End       285   
2                                   Spectre    206647   
3                     The Dark Knight Rises     49026   
4                               John Carter     49529   

                                               cast  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]   
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]   
3      [Christian Bale, Michael Caine, Gary Oldman]   
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [208]:
movies_1["crew"][0]

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

In [209]:
#Extracting only the directors from the crew column as people tend to usually search for movies based on their favourite directors

def fetch_director(txt):
    lst_2=[]
    
    for i in ast.literal_eval(txt):
        if i["job"]=="Director":
            lst_2.append(i["name"])
    
    return lst_2

In [210]:
movies_1["crew"]=movies_1["crew"].apply(fetch_director)


In [211]:
movies_1.head()

genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                      title  movie_id  \
0                                    Avatar     19995   
1  Pirates of the Caribbean: At World's End       285   
2                                   Spectre    206647   
3                     The Dark Knight Rises     49026   
4                               John Carter     49529   

                                               cast                 crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]      [James Cameron]  
1     [Johnny Depp, Orlando Bloom, Keira Knightley]     [Gore Verbinski]  
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]         [Sam Mendes]  
3      [Christian Bale, Michael Caine, Gary Oldman]  [Christopher Nolan]  
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]     [Andrew Stanton]

In [212]:
movies_1.isna().sum()

genres      0
keywords    0
overview    3
title       0
movie_id    0
cast        0
crew        0
dtype: int64

In [213]:
movies_1.isna().sum()/len(movies_1)*100

genres      0.000000
keywords    0.000000
overview    0.062383
title       0.000000
movie_id    0.000000
cast        0.000000
crew        0.000000
dtype: float64

In [214]:
movies_1.shape

(4809, 7)

In [215]:
#Since overview has very less number of np.nan values we can remove those rows

movies_1.dropna(axis=0,inplace=True)

In [216]:
movies_1.shape

(4806, 7)

In [217]:
movies_1["overview"][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

In [218]:
#Now concating the cols genres	keywords	overview	cast	crew under one column context

movies_1["overview"] = movies_1["overview"].apply(lambda x: x.split(' '))
movies_1.head()

genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                      title  movie_id  \
0                                    Avatar     19995   
1  Pirates of the Caribbean: At World's End       285   
2                                   Spectre    206647   
3                     The Dark Knight Rises     49026   
4                               John Carter     49529   

                                               cast                 crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]      [James Cameron]  
1     [Johnny Depp, Orlando Bloom, Keira Knightley]     [Gore Verbinski]  
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]         [Sam Mendes]  
3      [Christian Bale, Michael Caine, Gary Oldman]  [Christopher Nolan]  
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]     [Andrew Stanton]

In [219]:
movies_1["context"] = movies_1["genres"] + movies_1["keywords"] + movies_1["overview"] + movies_1["cast"] + movies_1["crew"]
movies_1.head()

genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                      title  movie_id  \
0                                    Avatar     19995   
1  Pirates of the Caribbean: At World's End       285   
2                                   Spectre    206647   
3                     The Dark Knight Rises     49026   
4                               John Carter     49529   

                                               cast                 crew  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]      [James Cameron]   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]     [Gore Verbinski]   
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]         [Sam Mendes]   
3      [Christian Bale, Michael Caine, Gary Oldman]  [Christopher Nolan]   
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]     [Andrew Stanton]   

                                             context  
0  [Action, Adventure, Fantasy, Science Fiction, ...  
1  [Adventure, Fantasy, Action, ocean, drug abuse...  
2  [Action, Adventure, Crime, spy, based on novel...  
3  [Action, Crime, Drama, Thriller, dc comics, cr...  
4  [Action, Adventure, Science Fiction, based on ...

In [220]:
movies_1["context"][0]

['Action',
 'Adventure',
 'Fantasy',
 'Science Fiction',
 'culture clash',
 'future',
 'space war',
 'space colony',
 'society',
 'space travel',
 'futuristic',
 'romance',
 'space',
 'alien',
 'tribe',
 'alien planet',
 'cgi',
 'marine',
 'soldier',
 'battle',
 'love affair',
 'anti war',
 'power relations',
 'mind and soul',
 '3d',
 'In',
 'the',
 '22nd',
 'century,',
 'a',
 'paraplegic',
 'Marine',
 'is',
 'dispatched',
 'to',
 'the',
 'moon',
 'Pandora',
 'on',
 'a',
 'unique',
 'mission,',
 'but',
 'becomes',
 'torn',
 'between',
 'following',
 'orders',
 'and',
 'protecting',
 'an',
 'alien',
 'civilization.',
 'Sam Worthington',
 'Zoe Saldana',
 'Sigourney Weaver',
 'James Cameron']

In [221]:
#Now removing the spaces in each word such that it will be easier for tokeninzation
#For e.g Sam Worthington and Sam Mendes will have thesame first name and multiple tokens will be created for Sam hence 
#inorder to avoid it we remove spaces ---SamWorthington will get 1 token and SamMendes will get 1 token instead of two each

def rem_space(txt):
    
    lst_replace=[]
    
    for i in txt:
        lst_replace.append(i.replace(" ",""))
        
    return lst_replace

In [222]:
movies_1["context"] = movies_1["context"].apply(rem_space)

In [223]:
movies_1.head()

genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                      title  movie_id  \
0                                    Avatar     19995   
1  Pirates of the Caribbean: At World's End       285   
2                                   Spectre    206647   
3                     The Dark Knight Rises     49026   
4                               John Carter     49529   

                                               cast                 crew  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]      [James Cameron]   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]     [Gore Verbinski]   
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]         [Sam Mendes]   
3      [Christian Bale, Michael Caine, Gary Oldman]  [Christopher Nolan]   
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]     [Andrew Stanton]   

                                             context  
0  [Action, Adventure, Fantasy, ScienceFiction, c...  
1  [Adventure, Fantasy, Action, ocean, drugabuse,...  
2  [Action, Adventure, Crime, spy, basedonnovel, ...  
3  [Action, Crime, Drama, Thriller, dccomics, cri...  
4  [Action, Adventure, ScienceFiction, basedonnov...

In [224]:
#Removing the columns joined

movies_1.drop(columns=["genres","keywords","overview","cast","crew"], axis=1,inplace=True)

movies_1.head(20)

title  movie_id  \
0                                        Avatar     19995   
1      Pirates of the Caribbean: At World's End       285   
2                                       Spectre    206647   
3                         The Dark Knight Rises     49026   
4                                   John Carter     49529   
5                                  Spider-Man 3       559   
6                                       Tangled     38757   
7                       Avengers: Age of Ultron     99861   
8        Harry Potter and the Half-Blood Prince       767   
9            Batman v Superman: Dawn of Justice    209112   
10                             Superman Returns      1452   
11                            Quantum of Solace     10764   
12   Pirates of the Caribbean: Dead Man's Chest        58   
13                              The Lone Ranger     57201   
14                                 Man of Steel     49521   
15     The Chronicles of Narnia: Prince Caspian      2454   
16                                 The Avengers     24428   
17  Pirates of the Caribbean: On Stranger Tides      1865   
18                               Men in Black 3     41154   
19    The Hobbit: The Battle of the Five Armies    122917   

                                              context  
0   [Action, Adventure, Fantasy, ScienceFiction, c...  
1   [Adventure, Fantasy, Action, ocean, drugabuse,...  
2   [Action, Adventure, Crime, spy, basedonnovel, ...  
3   [Action, Crime, Drama, Thriller, dccomics, cri...  
4   [Action, Adventure, ScienceFiction, basedonnov...  
5   [Fantasy, Action, Adventure, dualidentity, amn...  
6   [Animation, Family, hostage, magic, horse, fai...  
7   [Action, Adventure, ScienceFiction, marvelcomi...  
8   [Adventure, Fantasy, Family, witch, magic, bro...  
9   [Action, Adventure, Fantasy, dccomics, vigilan...  
10  [Adventure, Fantasy, Action, ScienceFiction, s...  
11  [Adventure, Action, Thriller, Crime, killing, ...  
12  [Adventure, Fantasy, Action, witch, fortunetel...  
13  [Action, Adventure, Western, texas, horse, sur...  
14  [Action, Adventure, Fantasy, ScienceFiction, s...  
15  [Adventure, Family, Fantasy, basedonnovel, fic...  
16  [ScienceFiction, Action, Adventure, newyork, s...  
17  [Adventure, Action, Fantasy, sea, captain, mut...  
18  [Action, Comedy, ScienceFiction, timetravel, t...  
19  [Action, Adventure, Fantasy, corruption, elves...

In [225]:
#Now converting the list under context to a string of words so that we can tokenize them

movies_1["context"] = movies_1["context"].apply(lambda x: " ".join(x))

movies_1["context"][0]

'Action Adventure Fantasy ScienceFiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. SamWorthington ZoeSaldana SigourneyWeaver JamesCameron'

# USING BAG OF WORDS (BOW) FOR TOKENIZATION

In [226]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(stop_words="english")
#CountVectorizer is a tool used for converting text data into a numerical format that machine learning algorithms can understand.

#max_features=5000: This parameter specifies the maximum number of features (unique words or tokens in vocab) that will be used to 
#create the numerical representations of the text data. In this case, the CountVectorizer will only consider the top 
#5000 most frequent words/tokens in the entire text data to be considered as vocab.

In [227]:
tokenized = cv.fit_transform(movies_1["context"])

In [228]:
tokenized

<4806x35547 sparse matrix of type '<class 'numpy.int64'>'
	with 190072 stored elements in Compressed Sparse Row format>

In [229]:
tokenized= tokenized.toarray()

In [230]:
tokenized

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [231]:
tokenized.shape

(4806, 35547)

Here 35547 is the no. of vocabulary (Which is set of unique words). You can set the max features to 5000 which is a parameter in countvectorize but i chose not to

# BUILDING RECOMMENDATION SYSTEM USING COSINE SIMILARITY

In [232]:
from sklearn.metrics.pairwise import cosine_similarity

In [233]:
c_sim = cosine_similarity(tokenized)
c_sim

array([[1.        , 0.07024394, 0.04622502, ..., 0.01848053, 0.0202113 ,
        0.        ],
       [0.07024394, 1.        , 0.05195243, ..., 0.02077032, 0.        ,
        0.        ],
       [0.04622502, 0.05195243, 1.        , ..., 0.02050231, 0.        ,
        0.        ],
       ...,
       [0.01848053, 0.02077032, 0.02050231, ..., 1.        , 0.05378625,
        0.0342193 ],
       [0.0202113 , 0.        , 0.        , ..., 0.05378625, 1.        ,
        0.03742406],
       [0.        , 0.        , 0.        , ..., 0.0342193 , 0.03742406,
        1.        ]])

The above matrix tells us how each row is similar to itself that is why 1 and also with other rows. c_sim is symmetric matrix with equal no. of rows and columns i.e 4806 rows and 4806 columns as each column is campared in terms of similarity with every other column w.r.t tokens/words

In [234]:
#Building a recommendation system based on the tokenized context

movies_1[movies_1["title"]=="Avatar"].index[0]
#Gives the index of that movie

0

In [235]:
movies_1[movies_1["title"]=="Pirates of the Caribbean: At World's End"].index[0]

1

In [236]:
#Defining a recommendation function

def recommend (movie):
    index = movies_1[movies_1["title"]==movie].index[0]
    similar_movies = sorted(list(enumerate(c_sim[index])), reverse=True, key=lambda x:x[1])
    
    #index = gives the index of the movie entered
    #sorted(list(enumerate(c_sim[index])) = gives the similarity of that index and then enumerate will give a tuple of (index,similarity value of each col)
    #Then put in a list and sort it in descending order based on the simalarity of each col.
    #similar_movies = sorted list of tuples (index, similarities) of the specified movie index in descending order
    
    for i in similar_movies[1:11]: #because the first highest similarity i.e 1 will be the similarity with the column itseld
        print(movies_1.iloc[i[0]]["title"])

In [237]:
recommend("Avatar")

Titan A.E.
Aliens vs Predator: Requiem
Aliens
Battle: Los Angeles
Independence Day
Predators
Lifeforce
Ender's Game
Small Soldiers
Edge of Tomorrow


In [238]:
recommend("The Dark Knight Rises")

The Dark Knight
Batman Begins
Batman Returns
Batman
Batman
Batman Forever
Batman & Robin
Batman v Superman: Dawn of Justice
Slow Burn
Batman: The Dark Knight Returns, Part 2


In [239]:
recommend("Men in Black 3")

The Helix... Loaded
Timecop
Men in Black
Men in Black II
In Time
Project Almanac
Rush Hour 2
Kate & Leopold
Looper
X-Men: Days of Future Past
